### PR TITLE
LR Sinkhorn improvements

### PR DESCRIPTION
**Pull request by meyerscetbon**
Here are the implementations of the update for the lr-sinkhorn algorithm:
- New init using  kmeans
- Rescaling of the gamma
- New stopping criterion 
- New implementation for the update of the cost matrices

### CODE REVIEW

File: ott/core/sinkhorn_lr.py [hunk 1]
```diff
@@ -338,15 +426,34 @@ def lr_costs(
...
+    grad_g = jnp.where(
+        self.epsilon != 0., grad_g + self.epsilon * jnp.log(state.g), grad_g
+    )
+    if self.gamma_init == "rescale":
+      norm_g = jnp.max(jnp.abs(grad_g)) ** 2
+    if self.gamma_init == "rescale":
+      self.gamma = self.gamma / max(norm_q, norm_r, norm_g)
     h = diag_qcr / state.g ** 2 - (self.epsilon -
                                    1 / self.gamma) * jnp.log(state.g)
+    c_q = grad_q - (1 / self.gamma) * jnp.log(state.q)
```
Comment (michalk8) [inline ott/core/sinkhorn_lr.py:454]: Would assign `log(state.{q,r,g})` to variables, since they can be reused when computed `grad_{q,r,g}`

File: ott/core/sinkhorn_lr.py [hunk 2]
```diff
@@ -338,15 +426,34 @@ def lr_costs(
...
         state.q * ot_prob.geom.apply_cost(state.r, axis=1), axis=0
     )
+    grad_g = -diag_qcr / state.g ** 2
+    grad_g = jnp.where(
+        self.epsilon != 0., grad_g + self.epsilon * jnp.log(state.g), grad_g
+    )
+    if self.gamma_init == "rescale":
+      norm_g = jnp.max(jnp.abs(grad_g)) ** 2
+    if self.gamma_init == "rescale":
+      self.gamma = self.gamma / max(norm_q, norm_r, norm_g)
```
Comment (michalk8) [inline ott/core/sinkhorn_lr.py:451]: jnp.max

File: ott/core/sinkhorn_lr.py [hunk 3]
```diff
@@ -30,12 +31,24 @@ class LRSinkhornState(NamedTuple):
   q: Optional[jnp.ndarray] = None
   r: Optional[jnp.ndarray] = None
   g: Optional[jnp.ndarray] = None
+  q_prev: Optional[jnp.ndarray] = None
```
Comment (michalk8) [inline ott/core/sinkhorn_lr.py:34]: Feel like the state is too large, would suggest the following:
- remove `{q,r,g}_prev`, have only `compute_criterion` in `LRSinkhornState` and refactor it as:
  ```python
  def compute_criterion(self, previous_state: "LRSinkhornState") -> float:
    err_1 = ...
    ...
    return err_1 + err_2 + err_3
  ```
- make `criterion: Optional[float] = None` -> `criterions: Optional[jnp.ndarray]` and store the value at each iteration - might be useful to plot the evolution in the future

File: ott/core/sinkhorn_lr.py [hunk 4]
```diff
@@ -496,22 +607,39 @@ def one_iteration(
     else:
       state = self.kernel_step(ot_prob, state, iteration)
 
+    # compute the criterion
+    criterion = state.compute_crit()
```
Comment (michalk8) [inline ott/core/sinkhorn_lr.py:611]: See my comment re `LRSinkhornState` refactoring. Here I imagine it would be like:
```python
...
new_state = self.lse_step(ot_prob, state, iteration)
criterion = new_state.compute_criterion(old_state)
criterions = state.criterions.at[iteration // self.inner_iterations].set(criterion)
...

```

**Review by michalk8 — CHANGES_REQUESTED** [review]
I had a look the the notebook that causes the conflict, there are no new changes.
Could you please update the branch as `git merge master -X theirs`?

File: ott/core/sinkhorn_lr.py [hunk 5]
```diff
@@ -114,20 +118,11 @@ def solution_error(
...
-  err_2 = ((1 / gamma) ** 2) * (kl(r, r_prev) + kl(r_prev, r))
-  err_3 = ((1 / gamma) ** 2) * (kl(g, g_prev) + kl(g_prev, g))
-  criterion = err_1 + err_2 + err_3
-  return criterion
-
-
-def kl(q1, q2):
-  ratio = jnp.log(q1) - jnp.log(q2)
-  return jnp.sum(q1 * ratio)
+def kl(q1, q2, clipping_value=1e-8):
```
Comment (michalk8) [inline ott/core/sinkhorn_lr.py:121]: I like this implementation better than the one I in `quad_problems.py`, would consider re-using it there as well.

File: ott/core/sinkhorn_lr.py [hunk 6]
```diff
@@ -383,6 +380,7 @@ def __call__(
         init_r = ot_sink_y.matrix
     else:
       raise NotImplementedError(self.init_type)
+    self.is_entropic = (self.epsilon != 0.0)
```
Comment (michalk8) [inline ott/core/sinkhorn_lr.py:383]: I had in mind
```python
@property
def is_entropy(self) -> bool:
  return self.epsilon != 0.0
```
since it's read-only + we can write docs for this

File: ott/core/sinkhorn_lr.py [hunk 7]
```diff
@@ -392,12 +390,10 @@ def norm_error(self) -> Tuple[int]:
 
   def _converged(self, state: LRSinkhornState, iteration: int) -> bool:
     criterion, count_escape, i, tol = state.criterion, state.count_escape, iteration, self.threshold
-    if i >= 2:
-      if criterion > tol / 1e-1:
+    if jnp.logical_not(i < 2):
```
Comment (michalk8) [inline ott/core/sinkhorn_lr.py:393]: I don't think this will still work (locally tests fail for different reason); the way it's not can easily be rewritten purely using `jax.logical_{and,or,not}` without the Python `if`s.

File: ott/core/sinkhorn_lr.py [hunk 8]
```diff
@@ -426,34 +422,27 @@ def lr_costs(
...
+    if self.gamma_rescale:
       norm_g = jnp.max(jnp.abs(grad_g)) ** 2
-    if self.gamma_init == "rescale":
-      self.gamma = self.gamma / max(norm_q, norm_r, norm_g)
-    h = diag_qcr / state.g ** 2 - (self.epsilon -
-                                   1 / self.gamma) * jnp.log(state.g)
-    c_q = grad_q - (1 / self.gamma) * jnp.log(state.q)
-    c_r = grad_r - (1 / self.gamma) * jnp.log(state.r)
-    h = -grad_g + (1 / self.gamma) * jnp.log(state.g)
+    if self.gamma_rescale:
```
Comment (michalk8) [inline ott/core/sinkhorn_lr.py:441]: I'd sligthly prefer to compute all the gradient norms in this `if` statement, it's more readable.

File: ott/core/sinkhorn_lr.py [hunk 9]
```diff
@@ -602,21 +591,32 @@ def one_iteration(
...
     else:
       state = self.kernel_step(ot_prob, state, iteration)
 
     # compute the criterion
-    criterion = state.compute_crit()
+    criterion = state.compute_criterion(previous_state)
+
+    # compute count_escape
+    tol = self.threshold
+    if iteration >= 2:
```
Comment (michalk8) [inline ott/core/sinkhorn_lr.py:605]: Same issue as before, `iteration` can be a tracer value;
```python
count_escape = state.count_escape + jnp.logical_and(iteration >= 2, criterion <= tol * 10.)
``` 
should work.

File: ott/core/sinkhorn_lr.py [hunk 10]
```diff
@@ -392,12 +390,10 @@ def norm_error(self) -> Tuple[int]:
 
   def _converged(self, state: LRSinkhornState, iteration: int) -> bool:
     criterion, count_escape, i, tol = state.criterion, state.count_escape, iteration, self.threshold
```
Comment (michalk8) [inline ott/core/sinkhorn_lr.py:392]: There's no attribute `.criterion`, only `.criterions`.

Comment (michalk8) [reply ott/core/sinkhorn_lr.py:392]: Same goes for the `_diverged(self, state: LRSinkhornState, iteration: int) -> bool:` function.
`

File: ott/core/sinkhorn_lr.py (hunk 5 above)
Comment (michalk8) [reply ott/core/sinkhorn_lr.py:121]: Also would add typehints for this.

**Review by michalk8 — CHANGES_REQUESTED** [review]
I got circular import error because of the `kmeans`, because it's placed in tools. Making it a relative import in `LRSinkhorn.__call__` solved this.

**Comment by michalk8** [comment]
@meyerscetbon @marcocuturi you can both review it now, there some TODOs left for me, but not critical

File: ott/core/sinkhorn_lr.py [hunk 11]
```diff
@@ -519,14 +582,48 @@ def output_from_state(
...
+        q=state.q,
+        r=state.r,
+        g=state.g,
+        ot_prob=ot_prob,
+        costs=state.costs,
+        criterions=state.criterions,
+    )
+
+  def _converged(self, state: LRSinkhornState, iteration: int) -> bool:
+
```
Comment (marcocuturi) [inline ott/core/sinkhorn_lr.py:594]: maybe worth writing a short comment here outlining main idea?

File: ott/core/sinkhorn_lr.py [hunk 12]
```diff
@@ -567,6 +663,18 @@ def make(
...
       jit=jit,
-      rng_key=rng_key,
       kwargs_dys=kwargs_dys
   )
+
+
+# TODO(michalk8): move to math utils
+def kl(q1: jnp.ndarray, q2: jnp.ndarray) -> float:
+  res_1 = -jsp.special.entr(q1)
+  res_2 = q1 * safe_log(q2)
```
Comment (marcocuturi) [inline ott/core/sinkhorn_lr.py:673]: have you considered 

https://jax.readthedocs.io/en/latest/_autosummary/jax.scipy.special.xlogy.html

it's also safe (as long as x ~ y ~ 0)

File: tests/core/sinkhorn_lr_test.py [hunk 13]
```diff
@@ -95,22 +108,26 @@ def test_euclidean_point_cloud(self, use_lrcgeom: bool, init_type: str):
 
     # Ensure cost is higher when using high entropy.
     # (Note that for small entropy regularizers, this can be the opposite
-    # due to non-convexity of problem and benefit of adding regularizer.
+    # due to non-convexity of problem and benefit of adding regularizer
```
Comment (marcocuturi) [inline tests/core/sinkhorn_lr_test.py:111]: add closing parenthesis!

File: tests/core/sinkhorn_test.py [hunk 14]
```diff
@@ -13,7 +13,7 @@
 # limitations under the License.
 
 # Lint as: python3
-"""Tests for the Policy."""
+"""Tests for Sinkhorn."""
```
Comment (marcocuturi) [inline tests/core/sinkhorn_test.py:16]: thanks for this catch :)

**Review by marcocuturi — APPROVED** [review]
Thanks a lot for this PR, it looks like we have a really great LR implementation now.

File: ott/core/sinkhorn_lr.py (hunk 12 above)
Comment (michalk8) [reply ott/core/sinkhorn_lr.py:673]: It's only safe in `x ~ 0`, not `y ~ 0`:

```python
import jax.numpy as jnp
import ott

x = jnp.ones(3)
y = jnp.ones(3).at[0].set(0) / 2

jax.scipy.special.xlogy(x, y)
# DeviceArray([      -inf, -0.6931472, -0.6931472], dtype=float32)

x * ott.core.sinkhorn_lr.safe_log(y)
# DeviceArray([-87.33655  ,  -0.6931472,  -0.6931472], dtype=float32)
```

File: ott/core/sinkhorn_lr.py [hunk 15]
```diff
@@ -60,12 +72,13 @@ def compute_reg_ot_cost(
...
   g = jax.lax.stop_gradient(g) if use_danskin else g
-  return jnp.sum(ot_prob.geom.apply_cost(r, axis=1) * q * (1.0 / g)[None, :])
+  # TODO(michalk8): clip `1. / g` here? or during iterations?
+  return jnp.sum(ot_prob.geom.apply_cost(r, axis=1) * q * (1. / g)[None, :])
 
 
 def solution_error(
     q: jnp.ndarray, r: jnp.ndarray, ot_prob: linear_problems.LinearProblem,
-    norm_error: jnp.ndarray, lse_mode: bool
+    norm_error: Tuple[int, ...], lse_mode: bool
```
[the remark this answers is not in the document]
Comment (michalk8) [reply ott/core/sinkhorn_lr.py:80]: It's converted to an array in this function, so now the types are correct. But I agree this should be refactored in the future.

File: ott/core/sinkhorn_lr.py (hunk 11 above)
Comment (michalk8) [reply ott/core/sinkhorn_lr.py:593]: Done.

File: ott/core/sinkhorn_lr.py [hunk 16]
```diff
@@ -163,29 +173,32 @@ def converged(self) -> bool:
...
+    return jnp.sum(self.q * other_geom.apply_cost(self.r, axis=1) * self._inv_g)
 
+  # TODO(michalk8): when refactoring the API, use a property
   def transport_mass(self) -> float:
     """Sum of transport matrix."""
     return self.marginal(0).sum()
 
+  @property
+  def _inv_g(self) -> jnp.ndarray:
+    # TODO(michalk8): clip?
```
[the remark this answers is not in the document]
Comment (michalk8) [reply ott/core/sinkhorn_lr.py]: Opted for not clipping in the end, `safe_log` seems to be fine, no NaNs. If we encounter them in the future, I will debug where they come from and possibly add this.

File: ott/core/sinkhorn_lr.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (marcocuturi) [reply ott/core/sinkhorn_lr.py:80]: feels like the conversation should happen higher up, to have a signature with jnps only

[the remark this answers is not in the document]
Comment (michalk8) [reply ott/core/sinkhorn_lr.py]: Done, changed to convert to an array where the property is defined, also in Sinkhorn.
